### PR TITLE
uttmm-24-53_29 & uttmm-56 (tables for resources)

### PR DESCRIPTION
**Pull request by RishkoPavel**
Implemented:
- back-end pagination for general resources
- front-end pagination for general resources
- back-end for resources for specific project
- front-end for resources for specific project

### CODE REVIEW

File: Angular/src/app/general-resources/general-resources-tables/general-resources-inner-table/general-resources-inner-table.component.html [hunk 1]
```diff
@@ -1,32 +1,29 @@
-<div class="mat-elevation-z8">
-    <table mat-table [dataSource]='dataSource'>
-  
-      
+<div class="mat-elevation-z8" *ngIf="showTable">
+    <table mat-table [dataSource]='dataSource'>
```
Comment (yrikturash) [inline Angular/src/app/general-resources/general-resources-tables/general-resources-inner-table/general-resources-inner-table.component.html]:
```suggestion
    <table mat-table [dataSource]="dataSource">
```

File: Angular/src/app/general-resources/general-resources-tables/general-resources-inner-table/general-resources-inner-table.component.html [hunk 2]
```diff
@@ -1,32 +1,29 @@
-<div class="mat-elevation-z8">
-    <table mat-table [dataSource]='dataSource'>
-  
-      
+<div class="mat-elevation-z8" *ngIf="showTable">
+    <table mat-table [dataSource]='dataSource'>      
+     
       <ng-container matColumnDef="id">
         <th mat-header-cell *matHeaderCellDef> ID  </th>
         <td mat-cell *matCellDef="let element; let i = index"> {{i+1}} </td>
```
Comment (yrikturash) [inline Angular/src/app/general-resources/general-resources-tables/general-resources-inner-table/general-resources-inner-table.component.html]:
```suggestion
        <td mat-cell *matCellDef="let i = index"> {{i+1}} </td>
```

File: Angular/src/app/general-resources/general-resources-tables/general-resources-inner-table/general-resources-inner-table.component.html [hunk 3]
```diff
@@ -1,32 +1,29 @@
...
         <td mat-cell *matCellDef="let element"> {{element.description}} </td>
       </ng-container>
-  
+        
       <tr mat-header-row *matHeaderRowDef="displayedColumns; sticky: true" ></tr>
       <tr mat-row *matRowDef="let row; columns: displayedColumns;"></tr>
     </table>  
-    <mat-paginator [length]="lengthDataArrForDataSource" [pageSize]="5" [pageSizeOptions]="[5, 10, 20]" showFirstLastButtons (page)="pageChanged($event)"></mat-paginator>
+    <mat-paginator [length]="this.allResourcesCount"  [pageSize]="this.columnsPerPage" 
+    [pageSizeOptions]="[5, 10, 20]" showFirstLastButtons (page)="pageChanged($event)"></mat-paginator>
```
Comment (yrikturash) [inline Angular/src/app/general-resources/general-resources-tables/general-resources-inner-table/general-resources-inner-table.component.html]: [5, 10, 20] - move to shared constants as pageSizeOptions variable

File: Angular/src/app/project-info/custom.datepipe.ts [hunk 4]
```diff
@@ -7,6 +7,6 @@ import { DatePipe } from '@angular/common';
   export class CustomDatePipe extends
                DatePipe implements PipeTransform {
     transform(value: any, args?: any): any {
-      return super.transform(value, 'y MM d H:mm');
+      return super.transform(value, 'y/MM/dd H:mm');
```
Comment (yrikturash) [inline Angular/src/app/project-info/custom.datepipe.ts]: most probably we will have the same dates format across the project - pls move it to the /shared/pipes and 'y/MM/dd H:mm' into Constants->DateFormats->LongDateString

File: Angular/src/app/project-info/project-info.component.ts [hunk 5]
```diff
@@ -33,20 +33,15 @@ export class ProjectInfoComponent implements OnInit, AfterViewInit {
...
+
   async getResourcesData() {
     if (!this.generateOnceResourcesTable) {
       const allResources = await this.resourceService.getAllResourcesByProjId(this.projectId);
       this.sortedResourcesByCategory = this.resourceService.sortAllResourcesByCategories(allResources);
+
     }
     this.generateOnceResourcesTable = true;
-    $('#resTabId').slideToggle('slow');
+    $('#resTabId').fadeToggle('slow');
```
Comment (yrikturash) [inline Angular/src/app/project-info/project-info.component.ts]: try to use angular animations:
https://www.freecodecamp.org/news/how-to-use-animation-with-angular-6-675b19bc3496/

File: Angular/src/app/project-info/resources-table-for-project/resources-internal-table/resources-internal-table.component.html [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    <ng-container matColumnDef="description">
+      <th mat-header-cell *matHeaderCellDef> Опис </th>
+      <td mat-cell *matCellDef="let element"> {{element.description}} </td>
+    </ng-container>
+
+    <tr mat-header-row *matHeaderRowDef="displayedColumns; sticky: true" ></tr>
+    <tr mat-row *matRowDef="let row; columns: displayedColumns;"></tr>
+  </table>  
+  <mat-paginator [length]="lengthDataArrForDataSource" [pageSize]="pageSize" [pageSizeOptions]="[5, 10, 20]" showFirstLastButtons></mat-paginator>
```
Comment (yrikturash) [inline Angular/src/app/project-info/resources-table-for-project/resources-internal-table/resources-internal-table.component.html]: remind me, what is the difference between this and general-resources-inner-table.component.html?

File: examples/MvcWeb/TheraLang/Services/ResourceService.cs [hunk 7]
```diff
@@ -52,25 +53,119 @@ public async Task UpdateResource(Resource resource, int updatetById)
...
+                                           Description = res.Description,
+                                           Url = res.Url,
+                                           File = res.File,
+                                           CategoryId = res.CategoryId,
+                                           ResourceCategory = res.ResourceCategory,
+                                           ResourceToProjects = res.ResourceToProjects,
+                                           UpdatedById = res.UpdatedById,
+                                           CreatedDateUtc = res.CreatedDateUtc,
+                                           UpdatedDateUtc = res.UpdatedDateUtc,
+                                       });
```
[the remark this answers is not in the document]
Comment (RishkoPavel) [reply examples/MvcWeb/TheraLang/Services/ResourceService.cs]: Nope, its not the same things. When i used LINQ join which just redeclare resource data, then sesponse looks like this:
image

But if i use only query like you offer me, then response will bee not contain required fields like  this: 
image

File: examples/MvcWeb/TheraLang/Services/ResourceService.cs [hunk 8]
```diff
@@ -52,25 +53,119 @@ public async Task UpdateResource(Resource resource, int updatetById)
...
+                    .Take(recordsPerPage).ToList();
+                return resourcesPerPages;
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Error when get all resources: ", ex);
+            }
+        }
+
+        public int CountResourcesByCategoryId(int categoryId)
```
[the remark this answers is not in the document]
Comment (RishkoPavel) [reply examples/MvcWeb/TheraLang/Services/ResourceService.cs]: Why I should not indicate that this method counts resources **ByCategoryId**?
How i will know what he doing, when i call he from another space

File: examples/MvcWeb/TheraLang/Services/ResourceService.cs [hunk 9]
```diff
@@ -52,25 +53,119 @@ public async Task UpdateResource(Resource resource, int updatetById)
...
+                                           ResourceCategory = res.ResourceCategory,
+                                           ResourceToProjects = res.ResourceToProjects,
+                                           UpdatedById = res.UpdatedById,
+                                           CreatedDateUtc = res.CreatedDateUtc,
+                                           UpdatedDateUtc = res.UpdatedDateUtc,
+                                       });
+
+                var resourcesPerPages = joinedResources.Skip((pageNumber - 1) * recordsPerPage)
+                    .Take(recordsPerPage).ToList();
+                return resourcesPerPages;
```
[the remark this answers is not in the document]
Comment (RishkoPavel) [reply examples/MvcWeb/TheraLang/Services/ResourceService.cs]: when i use query that you offer me, i get this:
image
But I need to get this:
image

File: Angular/src/app/project-info/resources-table-for-project/resources-internal-table/resources-internal-table.component.html (hunk 6 above)
Comment (RishkoPavel) [reply Angular/src/app/project-info/resources-table-for-project/resources-internal-table/resources-internal-table.component.html]: they use different data acquisition structures. one takes all the data at once and then uses it and the second takes the data through pagination

File: Angular/src/app/project-info/project-info.component.ts (hunk 5 above)
Comment (RishkoPavel) [reply Angular/src/app/project-info/project-info.component.ts]: i implemented angular animation for table but, іt did not work well, most likely it is not suitable for my table

File: Angular/src/app/project-info/resources-table-for-project/resources-internal-table/resources-internal-table.component.html (hunk 6 above)
Comment (RishkoPavel) [reply Angular/src/app/project-info/resources-table-for-project/resources-internal-table/resources-internal-table.component.html]: different way of receiving, processing and presenting data

File: Angular/src/app/project-info/project-info.component.ts (hunk 5 above)
Comment (RishkoPavel) [reply Angular/src/app/project-info/project-info.component.ts]: I tried to use this, but it does not work with my component.
image
image
image

Comment (RishkoPavel) [reply Angular/src/app/project-info/project-info.component.ts]: image

**Pull request merged**